server.go

```diff
@@ -65,21 +66,19 @@ func NewServer(config ServerConfig) (*Server, error) {
 	}
 
 	for _, cfg := range s.packetConnConfigs {
-		am, err := s.createAllocationManager(cfg.RelayAddressGenerator, cfg.PermissionHandler)
-		if err != nil {
-			return nil, fmt.Errorf("failed to create AllocationManager: %w", err)
+		if err := s.startReadLoop(cfg.RelayAddressGenerator, cfg.PermissionHandler, func(am *allocation.Manager) {
+			s.readLoop(cfg.PacketConn, am)
+		}); err != nil {
+			return nil, err
 		}
-
-		go s.readPacketConn(cfg, am)
 	}
 
 	for _, cfg := range s.listenerConfigs {
-		am, err := s.createAllocationManager(cfg.RelayAddressGenerator, cfg.PermissionHandler)
-		if err != nil {
-			return nil, fmt.Errorf("failed to create AllocationManager: %w", err)
+		if err := s.startReadLoop(cfg.RelayAddressGenerator, cfg.PermissionHandler, func(am *allocation.Manager) {
```

Comment:
I perfectly understand why the callback is necessary here but do we really need to complicate our lives that much just to save on calling `createAllocationManager` twice (once in the `PackerConnConfig` loop and then in the `ListenerConfig` loop)? My head hurts every time I try to understand what's going on here. I mean it's elegant and nice and we agreed to make it this way but somehow I feel that this level of generalization hurts readabiliy.

Just my 2 cents of course.

Reply:
I fully agree. This reduced readability. I've taken the DRY principle too far here.

I reverted it.